File: python/searching/binary_search.py
"""
Binary Search
-------------
Time:  O(log n)
Space: O(1) iterative, O(log n) recursive

Requires sorted input. Halves the search space each step.
I included both versions since interviews sometimes ask for one or the other.
"""


def binary_search(arr, target):
    """Iterative version — generally preferred to avoid stack overhead."""
    low, high = 0, len(arr) - 1

    while low <= high:
        mid = low + (high - low) // 2  # avoids overflow in other languages

        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1

    return -1


def binary_search_recursive(arr, target, low=0, high=None):
    if high is None:
        high = len(arr) - 1

    if low > high:
        return -1

    mid = low + (high - low) // 2

    if arr[mid] == target:
        return mid
    elif arr[mid] < target:
        return binary_search_recursive(arr, target, mid + 1, high)
    else:
        return binary_search_recursive(arr, target, low, mid - 1)

File: javascript/searching/binarySearch.js
/**
 * Binary Search
 * Time:  O(log n)
 * Space: O(1) iterative / O(log n) recursive
 *
 * Array must be sorted.
 */

function binarySearch(arr, target) {
  let low = 0;
  let high = arr.length - 1;

  while (low <= high) {
    const mid = low + Math.floor((high - low) / 2);

    if (arr[mid] === target) return mid;
    if (arr[mid] < target) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }

  return -1;
}

function binarySearchRecursive(arr, target, low = 0, high = arr.length - 1) {
  if (low > high) return -1;

  const mid = low + Math.floor((high - low) / 2);

  if (arr[mid] === target) return mid;
  if (arr[mid] < target) {
    return binarySearchRecursive(arr, target, mid + 1, high);
  }
  return binarySearchRecursive(arr, target, low, mid - 1);
}

module.exports = { binarySearch, binarySearchRecursive };
